pkg/client/apiutil/restmapper.go

```diff
@@ -162,25 +168,34 @@ func (m *mapper) addKnownGroupAndReload(groupName string, versions ...string) er
 	// If no specific versions are set by user, we will scan all available ones for the API group.
 	// This operation requires 2 requests: /api and /apis, but only once. For all subsequent calls
 	// this data will be taken from cache.
-	if len(versions) == 0 {
-		apiGroup, err := m.findAPIGroupByName(groupName)
+	//
+	// Calling this also tells us if the server supports aggregated discovery, which allows
+	// loading everything with two api calls which we assume is overall cheaper, hence we always
+	// do this once optimistically to check and keep doing it if its supported.
+	if len(versions) == 0 || !m.checkedIfSupportsAggregatedDiscovery.Load() {
+		apiGroup, didLookup, err := m.findAPIGroupByNameAndMaybeAggregatedDiscovery(groupName)
 		if err != nil {
 			return err
 		}
-		if apiGroup != nil {
+		if apiGroup != nil && len(versions) == 0 {
 			for _, version := range apiGroup.Versions {
 				versions = append(versions, version.Version)
 			}
 		}
-	}
-
-	m.mu.Lock()
-	defer m.mu.Unlock()
 
-	// Create or fetch group resources from cache.
-	groupResources := &restmapper.APIGroupResources{
-		Group:              metav1.APIGroup{Name: groupName},
-		VersionedResources: make(map[string][]metav1.APIResource),
+		// No need to do anything further if aggregatedDiscovery is supported and we did a lookup
+		if m.supportsAggregatedDiscovery.Load() && didLookup {
+			failedGroups := make(map[schema.GroupVersion]error)
+			for _, version := range versions {
+				if m.knownGroups[groupName] == nil || m.knownGroups[groupName].VersionedResources[version] == nil {
+					failedGroups[schema.GroupVersion{Group: groupName, Version: version}] = &meta.NoResourceMatchError{}
```

Comment:
Maybe worth a comment that we'll hit this case for:
* groupVersions that don't exist
* groupVersions that are stale (via aggregated discovery)

If I understand this correctly

Reply:
Although I think we hit the latter case only if the groupsVersions always were stale?

Reply:
Makes me wonder if we should drop stale groups from the cache or keep serving stale data. Probably the latter is better though

Reply:
Yeah I wasn't entirely sure what the right thing to do there is either. If I understand the kep correctly, stale just means that the CRD is not yet initialized which should rarely happen

Reply:
My take on the KEP was that stale (also?) happens if the kube-apiserver is not able to communicate with aggregated apiserves anymore

Reply:
Definiteily for initial sync, but probably also later if that happens

```
No, but if aggregated apiservers are present, the feature will attempt to contact and aggregate the data published from the aggregated apiserver on a set interval. If there is high error rate, stale data may be returned because the latest data was not able to be fetched from the aggregated apiserver.
```

Reply:
I think the alternative would be to differentiate between NoResourceMatchError and StaleGroupVersionError.

But maybe it's fine to surface always NoResourceMatchError basically StaleGroupVersionError means that the apiserver didn't fine a "fresh" version of the resource. Maybe NoResourceMatchError is close in enough even for that case?

Reply:
But should we return this as an error? After re-reading, I actually removed the deletion of stale groups, because I don't think that restmappings change that often, so having potentially stale informations is likely better than having no information. At the end of the day, controllers will retry anyways should there be a missing mapping and that triggers a new non-aggregated discovery call.

Reply:
Sounds reasonable. I'll take another look at the code

Reply:
Looks good!

Also it looks like failedGVs aren't even added to resources: https://github.com/kubernetes/client-go/blob/master/discovery/aggregated_discovery.go#L82-L83

(so the previous code was a no-op with the current aggregated discovery implementation)